Project creation, dependency locking, and multiple environments defined in conda-project.yml

AlbertDeFusco opened the pull request:
This PR addresses:
* #3
* #6
* #8
* #10
* #19 
* #20 
* #21
* #25

TL;DR: Do I need a `conda-project.yml` file? No. What can I do with a `conda-project.yml` file? Manage multiple environments and each environment with multiple source environment YAML files. Oh, and dependencies are automatically locked using conda-lock.

## If I already have an `environment.yml`

This PR adds automatic locking of package versions when running `conda project prepare` or by calling `conda project lock`. The locked dependencies are generated by [conda-lock](https://github.com/conda-incubator/conda-lock) in the file `default.conda-lock.yml`. Locking of dependencies completely ignores the user's channel settings in `~/.condarc` (issue #20) and will only use channels supplied by the environment.yml file.

If the user writes a minimal environment.yml file as shown below conda project will make two assumptions 1) all packages come from the `defaults` channel and 2) the dependencies will be locked for `win-64, linux-64, mac-64` and your current platform if it is not one of those three.

```yaml
name: project

dependencies:
  - python=3.8
```

`conda project lock` utilizes  the `--check-input-hash` feature of conda-lock. When you run `conda project lock` multiple times the lock file will only be updated if the environment.yml has changed. To force a re-lock use `conda project lock --force`.

## conda-project.yml file

An optional `conda-project.yml` file is defined that supports multiple conda environments per project and each environment can be built from multiple Conda environment YAML sources, which uses [conda-lock compound speicification](https://github.com/conda-incubator/conda-lock#compound-specification). For example:

```yaml
name: project-name

environments:
  main:
    - environment.yml
  development:
    - environment.yml
    - ../dev-extras.yml
```

Notice that environment YAML files are specified as relative to the location of the conda-project.yml file. Each key in `environments:` can be utilized in `conda project lock <env-name>` or `conda project prepare <env-name>`. These commands also accept `--all` to lock and prepare each of the defined environments. 

If no env name is supplied lock, prepare, and clean assume the *default environment* is the first environment listed in the `conda-project.yml` file.




## prepare
`conda project prepare` now forces the use of conda-lock. If a `.conda-lock.yml` file is not present it will be created by prepare with the above assumptions if necessary. If a `.conda-lock.yml` file is found but the locked platforms do not match your current platform it will raise an exception.

The live conda environment is built from a rendered lockfile (explicit type) for your current platform, similar to how `conda lock install` works.

## project creation

```
usage: conda-project create [-h] [--directory PROJECT_DIR] [-n NAME] [-c CHANNEL] [--platforms PLATFORMS] [--conda-configs CONDA_CONFIGS]
                            [--no-lock] [--prepare]
                            [dependencies [dependencies ...]]

Create a new project

positional arguments:
  dependencies          Packages to add to the environment.yml in MatchSpec format.

optional arguments:
  -h, --help            show this help message and exit
  --directory PROJECT_DIR
                        Project directory (defaults to current directory)
  -n NAME, --name NAME  Name for the project.
  -c CHANNEL, --channel CHANNEL
                        Additional channel to search for packages. The default channel is 'defaults'. Multiple channels are added with repeated
                        use of this argument.
  --platforms PLATFORMS
                        Comma separated list of platforms for which to lock dependencies. The default is win-64,linux-64,osx-64,osx-arm64
  --conda-configs CONDA_CONFIGS
                        Comma separated list of Conda configuration parameters to write into the .condarc file in the project directory. The
                        format for each config is key=value. For example --conda-configs experimental_solver=libmamba,channel_priority=strict
  --no-lock             Do no create the conda-lock.yml file
  --prepare             Create the local Conda environment for the current platform.
```

The purpose of `conda project create` is to provide a command like `conda create` that creates the `environment.yml`, `conda-project.yml` and `default.conda-lock.yml` files before installing the environment. The `--prepare` flag can be used to build the files and then install the environment.

The create command will always write `channels` and `platforms` into the environment.yml file. Here's a minimal example.

```
❯ conda project create python=3.8
Locking dependencies for default: done
Locked dependencies for win-64, osx-64, osx-arm64, linux-64 platforms
Project created at /Users/adefusco/Development/conda-incubator/conda-project/examples/new-project

❯ tree -a ./
./
├── .condarc
├── conda-project.yml
├── default.conda-lock.yml
└── environment.yml

❯ cat environment.yml
name: new-project
channels:
  - defaults
dependencies:
  - python=3.8
platforms:
  - win-64
  - osx-64
  - osx-arm64
  - linux-64

❯ cat conda-project.yml
name: new-project
environments:
  default:
    - environment.yml
```

## Python API

The Python API provides full support for the above workflows by creating a `CondaProject` object. `CondaProject` takes a single optional argument to supply the path to the project. The default value is the current working directory, `.` Every CondaProject has at least one Conda environment. 

A project directory containing only an `environment.yml` file will create a single enviroment of the name `default`, which can be locked or prepared. If multiple environments are defined in a `conda-project.yml` the `.environments` attribute provides dictionary-style syntax for each named environment. The first defined conda environment in the `conda-project.yml` is accessible as `project.default_environment`

```python
from conda_project import CondaProject

project = CondaProject()
project.default_environment.lock()
prefix = project.default_environment.prepare()

## alternative, use the name 'default'

project.environments['default'].lock()
prefix = project.environments['default'].prepare()
```

To create a new project directory the `CondaProject.create()` method follows the CLI arguments described above. Projects are automatically locked. See the docstring for `.create()` for more details.

```python
from conda_project import CondaProject

project = CondaProject.create(
  directory='new-project',
  dependencies=['python=3.8']
```

AlbertDeFusco committed:
utilize conda-lock

the lock command constructs a conda-lock.yml file in
the project directory from the environment.yml file.

if channels is not provided in the env file 'defaults' is
assumed

if platforms is not provided in the env file the current
platform (as identified by conda) is used plus the three
most common platforms linux-64, osx-64, and win-64

AlbertDeFusco committed:
a little refactor of calling conda-lock

this should help keep testing simple

AlbertDeFusco committed:
add verbose to cli

AlbertDeFusco committed:
new conda tests

AlbertDeFusco committed:
ensure condarc is utilized

AlbertDeFusco committed:
locking tests

AlbertDeFusco committed:
improve local condarc support

AlbertDeFusco committed:
full test coverage

AlbertDeFusco committed:
cli tests run in a real project directory

AlbertDeFusco committed:
spinner and logging

conda-lock does not provide any useful feedback
so a spinner has been added while it runs

AlbertDeFusco committed:
more locking tests

AlbertDeFusco committed:
maintain verbose

AlbertDeFusco committed:
check that current platform is supported

AlbertDeFusco committed:
project creation command

AlbertDeFusco committed:
full coverage

AlbertDeFusco committed:
depend on conda-lock

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

AlbertDeFusco committed:
Update test-environment.yml

AlbertDeFusco committed:
Fix windows tests

AlbertDeFusco committed:
isort and black are incompatible

see: https://ice-panda.medium.com/pre-commit-hooks-claim-files-are-modified-but-no-changes-are-detected-by-git-c43217f51d8a

AlbertDeFusco committed:
add ruamel.yaml dependency

AlbertDeFusco committed:
one more file

AlbertDeFusco committed:
swap logger.warning for warnings.warn

AlbertDeFusco committed:
add conda 4.13 to test matrix

AlbertDeFusco committed:
spinner

jlstevens commented:
The described behavior sounds good!

Personally, I would be more likely to create my `environment.yml` or `project.yml` and then run `conda project lock` but I can see how `conda project create` would be useful. 

I note that only a `environment.yml` is being created here: am I right in thinking a `project.yml` would also work? Don't you still need a `project.yml` to define commands or does `environment.yml` support this now? If not, I would expect the `conda project create` to also create a stub `project.yml` so I have somewhere to define commands. 

Even if everything can now be specified with an `environment.yml` (if so, I missed this!), my issue with this name is that it gives no indication that the current directory is a *project* (i.e. has commands). Most of the  `environment.yml` files in the world right now weren't written with the idea of a 'project' in mind unfortunately! If these modes are supported, shouldn't `conda project create` give you the option to create 1) just an `environment.yml` (as shown above) 2) a `project.yml` and `environment.yml` pair or 3) a self-contained `project.yml`?

Regarding platforms:

> the dependencies will be locked for win-64, linux-64, mac-64 and your current platform if it is not one of those three.

I think this is a good default and better than the old anaconda-project behavior which would refuse to lock without platforms defined (iirc). This default will cover the vast majority of cases and it makes sense to me that you only need to define the platforms if you want to specify something different.

The strategy of ensuring `prepare` uses `lock` also makes a lot of sense to me - this is how we can make sure locks don't go stale!

AlbertDeFusco commented:
Thanks! I haven't implemented project.yml or commands yet, but I see your point about making a project.yml. Right now I think the minimal project.yml file would define a conda env (not sure if I want to keep "env_specs") referencing the generated environment.yml file.

I don't have plans to extend environment.yml for things like commands.

AlbertDeFusco commented:
Right now a blank `.condarc` is not written if you don't specify `--conda-configs`, but I can do that for completeness

mattkram commented:
Sorry, late on the train here. I really like this. And I really like the idea of a blank `project.yml`. I would almost prefer that to have a key in it called something like `environment_file: environment.yml` to make it very explicit that this is a project, and which environment file to use. This would open up other use cases where different environments may compose from different environment files.

Also, wanted to throw out an alternative. How would folks feel about making `--prepare` default to True, and add instead a `--no-prepare` argument?

Finally, is `experimental_solver` a pass-through to conda itself? The format of that argument is kind of ugly next to the others and I'm wondering if it would warrant a `--solver` argument ...

AlbertDeFusco committed:
project file parsing, validation, and dumping

This includes conda-project.yml and environment.yml files.

AlbertDeFusco committed:
multiple environments and project file

environments defined in the project Yaml file support multiple sources

AlbertDeFusco committed:
revise project_directory_factory

AlbertDeFusco committed:
immutable environment models

AlbertDeFusco committed:
depend on pydantic

AlbertDeFusco committed:
tests for multiple envs and sources

added more cleaning steps

AlbertDeFusco committed:
add multiple env support to cli

lock, prepare, and clean accept an
optional environment name or --all

AlbertDeFusco committed:
full test coverage

AlbertDeFusco committed:
encode Path as_posix()

AlbertDeFusco committed:
refactor environment commands

lock, prepare, and clean functions now belong to each environemnt

AlbertDeFusco commented:
Thank you for your input. 

* Any empty .condarc is written by `conda project create` even if no explicit parameters are provided
* I've revised this PR to include support and creation of `conda-project.yml`, which is used to defined one or more named Conda environments. See the top of this PR for modified summary.

@mattkram , yes `experimental_solver` is a passthrough for the current version of Conda. At some point swappable solvers will no longer be experimental.

AlbertDeFusco committed:
always write .condarc using create

AlbertDeFusco committed:
revised test

AlbertDeFusco committed:
more detailed error messages

especially when locking

AlbertDeFusco committed:
handle problems with yml files

* strict enforcement of schema
* catch and raise validation errors
* raise on empty file

AlbertDeFusco committed:
capture more output from conda-lock

locking seems to only print to stderr on occassion.

AlbertDeFusco committed:
.gitignore in env prefix

AlbertDeFusco committed:
environment updates

AlbertDeFusco committed:
is_locked checks that the lock if up-to-date

by checking the computed hash of the environment sources
against the written has in the lockfile if it exists

AlbertDeFusco committed:
fix github action libmamba test

* can only be installed for conda >=4.12
* skip tests if libmamba not installed

AlbertDeFusco committed:
fix windows libmamba tests

AlbertDeFusco committed:
add is_prepared

is_prepared checks that lock file is up-to-date and that
a previously prepared env is consistent with the specs.

if the lock is outdated is_prepared returns false,
if the lock is correct, but the live
env is different it returns false

AlbertDeFusco committed:
docstrings

jlstevens commented on conda_project/cli/main.py:
```diff
@@ -40,12 +41,111 @@ def cli() -> ArgumentParser:
+        default=None,
+    )
+    p.add_argument(
+        "--no-lock", help="Do no create the conda-lock.yml file", action="store_true"
```
```suggestion
        "--no-lock", help="Do not create the conda-lock.yml file", action="store_true"
```

jlstevens commented on conda_project/cli/main.py:
```diff
@@ -40,12 +41,111 @@ def cli() -> ArgumentParser:
+    )
+    p.add_argument(
+        "dependencies",
+        help="Packages to add to the environment.yml in MatchSpec format.",
```
Wondering if this help string could be improved...'MatchSpec format' is meaningful to us but maybe a different description (or even a link?) might be more helpful to new users.

AlbertDeFusco replied:
I agree. There is an open issue on Conda to develop this resource 

https://github.com/conda/conda/issues/10491

I've added a minimal description to the help here.

jlstevens commented on conda_project/cli/main.py:
```diff
@@ -40,12 +41,111 @@ def cli() -> ArgumentParser:
+        "--channel",
+        help=(
+            "Additional channel to search for packages. The default channel is 'defaults'. "
+            "Multiple channels are added with repeated use of this argument."
```
Just wondering, does `conda-project` allow you to set the channel ordering list explicitly, including an explicit `defaults` as `anaconda-project` now does? If so, great!

AlbertDeFusco replied:
I think I understand your question, and yes. For example these two commands will produce different locks

```
conda project create -c defaults -c conda-forge python=3.9 sqlalchemy
```

```
conda project create -c conda-forge -c defaults python=3.9 sqlalchemy
```

AlbertDeFusco replied:
If you do `conda project create -c conda-forge python=3.9 sqlalchemy` the defaults channel will NOT be included.

That means this command will fail since `conda-token` is not in conda-forge

```
conda project create -c conda-forge python=3.9 conda-token
```

jlstevens replied:
Thanks for clarifying. I was thinking more of how `anaconda-project` can now get rid of the default of `defaults` by specifying `nodefaults`. However, it looks like `conda-project` is not taking this approach (which is fine, as it is explicit in a different way).

In addition to the locks being different, I assume your two `conda project create` commands above would list the channels in the specified order under the `channels` key in the `conda-project.yml`?

jlstevens commented:
Other than a few minor comments above, looks good!

Am I right in thinking that commands aren't yet supported? To me that is the one missing piece that I would need before starting to use `conda-project` in practice.

Happy to see this merged!

AlbertDeFusco committed:
Update conda_project/cli/main.py

Co-authored-by: Jean-Luc Stevens <jlrstevens@gmail.com>

AlbertDeFusco committed:
more dict-like methods for .environments

AlbertDeFusco committed:
conda project check

cli command to check for lockfile consistency

refactored lockfile checking and creation if out-of-date

prepare will print a message if the env and source/lockfile do not match

mattkram commented:
This isn't a blocker, but I have noticed an error when running the test locally. I believe it has to do with authentication and my system default of accessing the cloud repo vs the public one.

We may want to modify the conda settings during testing to ensure we only test locking against the public repo.

Error log below:

```
tests/test_project.py:225 (test_is_prepared_live_env_changed[.yml])
Traceback (most recent call last):
  File "/Users/mkramer/dev/conda-project/tests/test_project.py", line 238, in test_is_prepared_live_env_changed
    _ = call_conda(
  File "/Users/mkramer/dev/conda-project/conda_project/conda.py", line 45, in call_conda
    raise CondaProjectError(f"Failed to run:\n  {print_cmd}\n{proc.stderr.strip()}")
conda_project.exceptions.CondaProjectError: Failed to run:
  /usr/local/Caskroom/miniconda/base/bin/conda install -p /private/var/folders/79/hwkxhzw51ps11f3qzt10plmr0000gp/T/pytest-of-mkramer/pytest-281/test_is_prepared_live_env_chan0/envs/default requests -y
CondaHTTPError: HTTP 000 CONNECTION FAILED for url <https://repo.anaconda.cloud/repo/main/osx-64/repodata.json>
Elapsed: -

An HTTP error occurred when trying to retrieve this URL.
HTTP errors are often intermittent, and a simple retry will get you on your way.
'https://repo.anaconda.cloud/repo/main/osx-64'

AlbertDeFusco committed:
improve coverage

AlbertDeFusco committed:
fix a minor verbose print

mattkram commented on conda_project/project.py:
```diff
@@ -3,12 +3,337 @@
+        # To avoid this scenario lockfiles are written to a temporary location
+        # and copied back to the self.lockfile path if successful.
+        tempdir = Path(tempfile.mkdtemp())
+        lockfile = tempdir / self.lockfile.name
```
I think a future minor refactor could handle this w/ a context manager, but LGTM.

AlbertDeFusco replied:
Yes, let's do that in a separate PR: see #31

mattkram commented on conda_project/project.py:
```diff
@@ -3,12 +3,337 @@
+                if channel not in specified_channels:
+                    specified_channels.append(channel)
+
+        with redirect_stderr(StringIO()) as _:
```
On second thought, we may want to unify these various contexts into one, which would make the logic a lot cleaner.

```
with self.tmp_lockfile(self.lockfile) as fp:
   make_lock_files()...
```

This is what `atomicwrites` does I believe, or something similar, but we shouldn't take on a dependency for it.

If this sounds good, I'm happy to contribute a patch in another PR.

AlbertDeFusco replied:
Yes, let's do that in a separate PR: see #31

mattkram commented on conda_project/project.py:
```diff
@@ -27,69 +353,181 @@ class CondaProject:
-    def _find_environment_file(self) -> Path:
-        """Find an environment file in the project directory.
+    @classmethod
+    def create(
```
Please see my comment at caller.

AlbertDeFusco replied:
See #36

mattkram reviewed (commented):
Overall it looks great! I do have one substantive comment on the API for `CondaProject.create`, and a bunch of small things throughout. Many of them are nitpicks and we could manage them later, but this was a large PR so didn't want to lose my train of thought.

AlbertDeFusco commented on .pre-commit-config.yaml:
```diff
@@ -9,11 +9,6 @@ repos:
     hooks:
     -   id: end-of-file-fixer
     -   id: trailing-whitespace
--   repo: https://github.com/pycqa/isort
```
I was running into some problems with isort and black disagreeing and I could not commit. d3f75e26359808dd2c6b14c5db009cf2a77e0cb9 references this link, which helped me debug it. Is there a better approach?

https://ice-panda.medium.com/pre-commit-hooks-claim-files-are-modified-but-no-changes-are-detected-by-git-c43217f51d8a

mattkram replied:
Ahh, yes. We need to run `isort` with the `black` profile: https://pycqa.github.io/isort/docs/configuration/black_compatibility.html

AlbertDeFusco replied:
Thanks for the fix!

AlbertDeFusco commented on conda_project/cli/commands.py:
```diff
@@ -16,22 +16,89 @@ def handle_errors(func: Callable[[Namespace], Any]) -> Callable[[Namespace], int
-            func(args)
-            return 0
+            ret = func(args)
+            if ret:
+                return 0
+            else:
+                return 1
```
The problem is that this function's return is sent back through the CLI, where `0` means All is Well. I guess a more compact approach would be `int(not bool(ret))`

mattkram replied:
🤦 Yeah, I had that backwards. This is obviously a minor one but just caught my eye.

mattkram replied:
actually `int(not ret)` works too, but that's less explicit.

AlbertDeFusco committed:
pop the environ key, rather than del

Co-authored-by: Matt Kramer <matthew.robert.kramer@gmail.com>

AlbertDeFusco committed:
compare .keys() with set()

AlbertDeFusco commented on conda_project/project_file.py:
```diff
@@ -0,0 +1,73 @@
+    def only_pip_key_allowed(cls, v):
+        for item in v:
+            if isinstance(item, dict):
+                if not item.keys() == ["pip"]:
```
thanks!

AlbertDeFusco commented on tests/conftest.py:
```diff
@@ -4,39 +4,52 @@
         files = files or {}
         for fn, contents in files.items():
             path = tmp_path / fn
+            path.parent.mkdir(parents=True, exist_ok=True)
```
The reasoning was that `fn` may not just be a file name, but also a file name in nested subdirectories, which could be a bit too clever. In this test the project directory (where conda-project.yml lives) is a subdirectory of `tmp_path`. Line 47 ensures that the `project/` directory is created.

https://github.com/conda-incubator/conda-project/blob/9fb45bc18519fc5d9853828b36d37d19996cb542/tests/test_project.py#L889-L895

AlbertDeFusco commented on tests/test_project.py:
```diff
@@ -40,13 +183,82 @@ def test_prepare_no_dependencies(project_directory_factory):
+    assert project.default_environment.is_locked
+    assert not project.default_environment.is_prepared
+
+    _ = project.default_environment.prepare(force=True)
```
prepare() returns the path to the environment, referred to as the prefix. It seemed like something reasonable to return, but maybe it's not really necessary, you can get it from `project.default_environment.prefix` anyway.

The reason for `_ =` is remove linting messages in my IDE

AlbertDeFusco commented on tests/test_project.py:
```diff
@@ -1,37 +1,180 @@
+
+
+def test_project_create_new_directory(tmpdir, capsys):
+    project_directory = os.path.join(tmpdir, "new-project")
```
Oh! I didn't know about that fixture. Thanks!

mattkram replied:
I think it's pretty new. My new favorite 😁

AlbertDeFusco replied:
done

AlbertDeFusco commented on tests/test_project.py:
```diff
@@ -58,21 +270,703 @@ def test_prepare_and_clean(project_directory_factory):
+
+
+def test_lock_with_platforms(project_directory_factory):
+    env_yaml = """name: test
```
I agree

mattkram commented:
I just manually used this locally, and the one additional piece of feedback I have is that I think the filename of the lockfiles should be changed if possible.

Right now, we have `conda-project.yml` & `environment.yml` -> `default.conda-lock.yml`.

I would prefer if we could name them `conda-project-lock.default.yml` or similar so that, in the case of multiple environments, all of the lockfiles would be listed alphabetically next to each other. And ideally, also alphabetically directly below `conda-project.yml`.

AlbertDeFusco committed:
fixed incorrect gitignore

AlbertDeFusco commented on conda_project/project_file.py:
```diff
@@ -0,0 +1,73 @@
+
+from pydantic import BaseModel, validator, ValidationError
+from pathlib import Path
+from typing import List, Dict, Optional, Union, OrderedDict, TextIO
```
Sure, if you've got a hook let's use that.

AlbertDeFusco replied:
Let's do this in separate PR (see #35 ). Is this the correct hook?

https://github.com/mattkram/planingfsi/blob/a9b454e0e2bc61b2557ca87743adbf389c11bb51/.pre-commit-config.yaml#L14-L18

Would that also cover [pep604](https://peps.python.org/pep-0604/)?

AlbertDeFusco committed:
improve readability with triple quotes

AlbertDeFusco committed:
refactor _find_file

correctly raises an except if multiple options are found

mattkram commented on tests/test_cli.py:
```diff
@@ -61,12 +64,179 @@ def mocked_command(command, args):
+@pytest.mark.parametrize("command", ENVIRONMENT_COMMANDS)
+def test_command_with_environment_name(command, monkeypatch, project_directory_factory):
+    env1 = env2 = "dependencies: []\n"
+    project_yaml = f"""\
```
I don't want to harp on this more, but this is my new favorite way to format things like this I discovered this weekend: https://github.com/mattkram/rusty-mooring/blob/main/tests/test_config.py#L15

Just adding FYI.

AlbertDeFusco replied:
ok

AlbertDeFusco committed:
dedent

AlbertDeFusco committed:
fix isort black compatibility

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

AlbertDeFusco commented:
@mattkram , any chance you know what's up with the pre-commit check? I attempted to add and fix isort.

mattkram committed:
Fix key of isort config

mattkram committed:
Silence flake8 error: line break before binary operator

mattkram committed:
Move isort settings to setup.cfg to save a file

mattkram commented:
> @mattkram , any chance you know what's up with the pre-commit check? I attempted to add and fix isort.

Fixed it. I ended up moving the `isort` settings into `setup.cfg` to save a file. You just had the wrong section name (used the `pyproject.toml` one instead of `.isort.cfg` one).

AlbertDeFusco committed:
minimal package_spec description

AlbertDeFusco committed:
swap tmpdir for tmp_path

AlbertDeFusco commented on conda_project/cli/commands.py:
```diff
@@ -16,22 +16,89 @@ def handle_errors(func: Callable[[Namespace], Any]) -> Callable[[Namespace], int
 @handle_errors
-def prepare(args: Namespace) -> None:
+def create(args: Namespace) -> bool:
+    project = CondaProject.create(
```
This would require some interesting changes to `__init__` as well. I'll open an issue for this: #36
